### PR TITLE
Showcase > Add search in training page

### PR DESCRIPTION
**Pull request by NourJammal**
#  Showcase >Training page

Add search bar
Add Topic filters

https://smartjira.atlassian.net/browse/CFA-270

### CODE REVIEW

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs [hunk 1]
```diff
@@ -36,4 +36,30 @@ public static class TrainingQueryableExtensions
         var result = await query.PaginateAsync(pageNumber, pageSize, randomIds);
         return new PagedList<TrainingList>(result.PaginatedItems, new PageItem(pageNumber, pageSize), result.TotalCount);
     }
+    public static async Task<PagedList<TrainingList>> SearchPaginatedTrainingsAsync(this IQueryable<TrainingList> query,
```
Comment (maxime-poulain) [inline src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs:39]:
```suggestion

    public static async Task<PagedList<TrainingList>> SearchPaginatedTrainingsAsync(this IQueryable<TrainingList> query,
```

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs [hunk 2]
```diff
@@ -36,4 +36,30 @@ public static class TrainingQueryableExtensions
...
+    {
+        if (!string.IsNullOrEmpty(searchKeyWord))
+        {
+            query = query.Where(trainingList => trainingList.Title.Contains(searchKeyWord));
+        }
+
+        var paginationResult = await query.PaginateAsync(pageNumber, pageSize);
+        return new PagedList<TrainingList>(paginationResult.PaginatedItems, new PageItem(pageNumber, pageSize), paginationResult.TotalCount);
+    }
+    public static async Task<PagedList<TrainingList>> SearchPaginatedTrainingsByTopicAsync(this IQueryable<TrainingList> query,
```
Comment (maxime-poulain) [inline src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs:52]:
```suggestion

    public static async Task<PagedList<TrainingList>> SearchPaginatedTrainingsByTopicAsync(this IQueryable<TrainingList> query,
```

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Web/Services/Training/ITrainingService.cs [hunk 3]
```diff
@@ -8,4 +8,22 @@ public interface ITrainingService
...
+    public Task<PagedList<TrainingListViewModel>> SearchTrainingViewModelsAsync(string? searchKeyword, int currentPage, int pageSize);
+
+    /// <summary>
+    /// Searches trainings by a topic.
+    /// </summary>
+    /// <param name="searchTopicId">The topic that will be used for the search.</param>
+    /// <param name="currentPage">The current page for the pagination.</param>
+    /// <param name="pageSize">The number of item returned by the search.</param>
+    /// <returns>A task that represents the asynchronous operation. The task's result is a paginated list of <see cref="TrainingListViewModel"/>.</returns>
+    public Task<PagedList<TrainingListViewModel>> SearchTrainingByTopicViewModelsAsync(int? searchTopicId, int currentPage, int pageSize);
```
Comment (maxime-poulain) [inline src/Showcase/src/Smart.FA.Catalog.Showcase.Web/Services/Training/ITrainingService.cs:28]: Une meilleure signature serait
> SearchTrainingViewModelsByTopicAsync
> or
> SearchTrainingViewModelsByTopicIdAsync

De cette manière tu annonces clairement que ce sont des training view model qui sont retournés et pas des trainings cherché par des topic view models.

**Review by maxime-poulain — CHANGES_REQUESTED** [review]

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs [hunk 4]
```diff
@@ -36,4 +36,34 @@ public static class TrainingQueryableExtensions
...
+    }
+
+    public static async Task<PagedList<TrainingList>> SearchPaginatedTrainingsByTopicAsync(this IQueryable<TrainingList> query,
+        int? searchTopic,
+        int pageNumber,
+        int pageSize)
+    {
+        if (searchTopic != null)
+        {
+            query = query.Where(trainingList => trainingList.Topic == searchTopic);
```
Comment (LievenPraet) [inline src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs:63]: Same lowercase remark here

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Web/Pages/Training/TrainingList/TrainingList.cshtml.cs [hunk 5]
```diff
@@ -1,49 +1,43 @@
...
-            .Where(training => training.Status == TrainingStatusType.Published.Id)
-            .Select(training => training.Id)
-            .Distinct();
-        var totalTrainerTrainings = await trainingIdQuery.CountAsync();
-
-        var paginatedIds = trainingIdQuery.OrderBy(trainingId => trainingId).Skip(offset).Take(ItemsPerPage);
-        var trainingList = await _context.TrainingList.Where(training => paginatedIds.Contains(training.Id)).ToListAsync();
 
-        var trainingListViewModel = trainingList.ToTrainingListViewModels();
+        Trainings = await _trainingService.SearchTrainingViewModelsAsync(searchKeyword, CurrentPage, ItemsPerPage);
```
Comment (LievenPraet) [inline src/Showcase/src/Smart.FA.Catalog.Showcase.Web/Pages/Training/TrainingList/TrainingList.cshtml.cs:29]: Any reason to include ViewModel in the function name?
```suggestion
        Trainings = await _trainingService.SearchTrainingsAsync(searchKeyword, CurrentPage, ItemsPerPage);
```

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Web/Pages/Training/TrainingList/TrainingList.cshtml.cs [hunk 6]
```diff
@@ -1,49 +1,43 @@
...
-        var pageItem = new PageItem(CurrentPage, ItemsPerPage);
-        Trainings = new PagedList<TrainingListViewModel>(trainingListViewModel, pageItem, totalTrainerTrainings);
+    public async Task<ActionResult> OnGetSearchTopicAsync(int? id)
+    {
+        if (CurrentPage <= 0)
+        {
+            return RedirectToNotFound();
+        }
 
+        Trainings = await _trainingService.SearchTrainingViewModelsByTopicIdAsync(id, CurrentPage, ItemsPerPage);
```
Comment (LievenPraet) [inline src/Showcase/src/Smart.FA.Catalog.Showcase.Web/Pages/Training/TrainingList/TrainingList.cshtml.cs:40]: Same remark as above

**Review by LievenPraet — CHANGES_REQUESTED** [review]

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs [hunk 7]
```diff
@@ -36,4 +36,34 @@ public static class TrainingQueryableExtensions
...
+    public static async Task<PagedList<TrainingList>> SearchPaginatedTrainingsAsync(this IQueryable<TrainingList> query,
+        string? searchKeyWord,
+        int pageNumber,
+        int pageSize)
+    {
+        if (!string.IsNullOrEmpty(searchKeyWord))
+        {
+            query = query.Where(trainingList => trainingList.Title.Contains(searchKeyWord) ||
+                                                trainingList.Goal.Contains(searchKeyWord) ||
+                                                trainingList.Methodology.Contains(searchKeyWord));
```
[the remark this answers is not in the document]
Comment (maxime-poulain) [reply src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs:49]: It's a IQueryable therefore it's okay as sql is case insensitive.

File: src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs (hunk 4 above)
Comment (maxime-poulain) [reply src/Showcase/src/Smart.FA.Catalog.Showcase.Infrastructure/Data/Extensions/TrainingQueryableExtensions.cs:63]: This is an enumeration therefore, the value is an integer.

**Review by LievenPraet — APPROVED** [review]

**Review by maxime-poulain — APPROVED** [review]

**Pull request merged**